### PR TITLE
Implement Range::createContextualFragment

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11496)

### CODE REVIEW

File: components/script/dom/range.rs [hunk 1]
```diff
@@ -893,6 +895,52 @@ impl RangeMethods for Range {
...
+            if elem.local_name() == "html" && *elem.namespace() == ns!(html) {
+                Element::new(Atom::from("body"),
+                             ns!(html),
+                             None,
+                             &document)
+            } else {
+                elem.root_element()
+            }
+        } else {
+            element.unwrap().downcast::<Element>().unwrap().root_element()
```
Comment (GuillaumeGomez) [inline components/script/dom/range.rs]: This seems incorrect but cannot find another way to get the expected type.

File: components/script/dom/range.rs [hunk 2]
```diff
@@ -893,6 +895,52 @@ impl RangeMethods for Range {
...
+        };
+
+        // Step 3.
+        let fragment_node = element.upcast::<Node>().parse_fragment(fragment);
```
Comment (GuillaumeGomez) [inline components/script/dom/range.rs]: It seems strange that I have to upcast to get back a node in order to use this method.

**Comment by GuillaumeGomez** [comment]
Fixes #11111.

**Comment by nox** [comment]
Step 3 needs to be rewritten, and there should be tests for this.

-S-awaiting-review +S-needs-code-changes +S-needs-tests +C-needs-test 

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 3 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/script/dom/range.rs, line 913 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11496#-KIygC-VZnHvoffyMpyf:-KIygC-VZnHvoffyMpyg:824018497) ([raw file](https://github.com/servo/servo/blob/fa7de63f53e7cec6d607cd4e50cd563ca9934a26/components/script/dom/range.rs#L913)):*
> ```Rust
> 
>         // Step 2.
>         let elem_copy = element.clone().unwrap();
> ```

This is wrong to unwrap, given it could be `None` if `node` was a document or a document fragment.

---

*[components/script/dom/range.rs, line 916 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11496#-KIyheKnFKkVMi_wuMRk:-KIyheKnFKkVMi_wuMRl:-1705364883) ([raw file](https://github.com/servo/servo/blob/fa7de63f53e7cec6d607cd4e50cd563ca9934a26/components/script/dom/range.rs#L916)):*
> ```Rust
>         let document = elem_copy.downcast::<Document>().unwrap();
>         let element = if element.is_none() {
>             Element::new(Atom::from("body"),
> ```

`Element::new` would create an `Element` instead of the correct `HTMLBodyElement`. You want `Element::create`.

---

*[components/script/dom/range.rs, line 929 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11496#-KIygURivbIuQzAlzHn7:-KIygURjZZ4cfDfJVuAH:327056442) ([raw file](https://github.com/servo/servo/blob/fa7de63f53e7cec6d607cd4e50cd563ca9934a26/components/script/dom/range.rs#L929)):*
> ```Rust
>                              &document)
>             } else {
>                 elem.root_element()
> ```

There is no returning the root element in the spec AFAICT.

---

*[components/script/dom/range.rs, line 932 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11496#-KIxPAWX5BlBI9_6P5ab-r1-932:-KIygXJcNWrUg6GowsKc:457943303) ([raw file](https://github.com/servo/servo/blob/fa7de63f53e7cec6d607cd4e50cd563ca9934a26/components/script/dom/range.rs#L932)):*
<details><summary><i>Previously, GuillaumeGomez (Guillaume Gomez) wrote…</i></summary>
> This seems incorrect but cannot find another way to get the expected type.
</details>
So you just need to create a body element in a specific case when a set of conditions are met. You can do:

```rust
let should_create_body = element.map_or(true, |elem| {
    elem.local_name == &atom!("html") && elem.html_element_in_html_document()
});
let element = if should_create_body {
    Root::upcast(HTMLBodyElement::new(atom!("body"), None, &node.owner_doc()))
} else {
    element
};
```

---

*[components/script/dom/range.rs, line 936 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11496#-KIxPAWX5BlBI9_6P5ab-r1-936:-KIyioZ4p4G5FGNGeKDq:655209233) ([raw file](https://github.com/servo/servo/blob/fa7de63f53e7cec6d607cd4e50cd563ca9934a26/components/script/dom/range.rs#L936)):*
<details><summary><i>Previously, GuillaumeGomez (Guillaume Gomez) wrote…</i></summary>
> It seems strange that I have to upcast to get back a node in order to use this method.
</details>
Nope, that's because parsing a fragment is an operation on `Node`, we do that all the time.

---

*[components/script/dom/range.rs, line 936 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11496#-KIyjKlt2cGW4qK91qGe:-KIyjKlt2cGW4qK91qGf:-426332750) ([raw file](https://github.com/servo/servo/blob/fa7de63f53e7cec6d607cd4e50cd563ca9934a26/components/script/dom/range.rs#L936)):*
> ```Rust
> 
>         // Step 3.
>         let fragment_node = element.upcast::<Node>().parse_fragment(fragment);
> ```

Please do `try!()` around the whole call so that `fragment_node` is actually a `Root<Node>` and not a `Fallible<Root<Node>>`, since all the steps haven't been executed yet at this point.

---

*[components/script/dom/range.rs, line 939 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11496#-KIyiuzTTFRgNpDLQ2Pz:-KIyiuzTTFRgNpDLQ2Q-:-336800383) ([raw file](https://github.com/servo/servo/blob/fa7de63f53e7cec6d607cd4e50cd563ca9934a26/components/script/dom/range.rs#L939)):*
> ```Rust
> 
>         // Step 4.
>         // can be ignored
> ```

Please write a proper sentence as to why it can be ignored.

---

*[components/script/dom/range.rs, line 942 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11496#-KIyjYw6cFy3Pmikiyv4:-KIyjYw6cFy3Pmikiyv5:1210950415) ([raw file](https://github.com/servo/servo/blob/fa7de63f53e7cec6d607cd4e50cd563ca9934a26/components/script/dom/range.rs#L942)):*
> ```Rust
> 
>         // Step 5.
>         fragment_node
> ```

```rust
Ok(fragment_node)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11496#-:-KIyjmt7DC7DJXrTY_Yq:-1502131477)*

**Comment by nox** [comment]
Just a few nits, but it still needs tests.

-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[components/script/dom/range.rs, line 913 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11496#-KJ6k1YAXWRfpzW4fOu9:-KJ6k1YAXWRfpzW4fOuA:-1224520100) ([raw file](https://github.com/servo/servo/blob/29c0cee5830f87afad8eb3323f220aefb413122b/components/script/dom/range.rs#L913)):*
> ```Rust
> 
>         // Step 2.
>         let should_create_body = element.clone().map_or(true, |elem| {
> ```

Do `element.as_ref()` instead of `element.clone()`.

---

*[components/script/dom/range.rs, line 917 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11496#-KJ6jhuk-N-IRAn66euW:-KJ6jhuk-N-IRAn66euX:1564486226) ([raw file](https://github.com/servo/servo/blob/29c0cee5830f87afad8eb3323f220aefb413122b/components/script/dom/range.rs#L917)):*
> ```Rust
>             elem.local_name() == &atom!("html") && elem.html_element_in_html_document()
>         });
>         let element: Root<Node> = if should_create_body || element.is_none() {
> ```

You can do it this way:

```rust
let element = Root::upcast::<Node>(if should_create_body { ... });
```

This removes the double call to `Root::upcast` and the `Root<Node` type annotation, and also the `element.is_none()` test that is already covered by `element.map_or(true, ...)`.

---

*[components/script/dom/range.rs, line 927 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11496#-KJ6kabCraG_A2k7HIs6:-KJ6kabCraG_A2k7HIs7:-1947577860) ([raw file](https://github.com/servo/servo/blob/29c0cee5830f87afad8eb3323f220aefb413122b/components/script/dom/range.rs#L927)):*
> ```Rust
> 
>         // Step 4.
>         // Can be ignored because at the initialization are supposed to be unmarked as
> ```

What are supposed to be unmarked as "already started"? There lack a subject in that sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11496#-:-KJ6khpIuHpzml_PB5a5:-1961140452)*

**Comment by nox** [comment]
-S-awaiting-review -S-needs-tests 

Seems like there are tests already in WPT. Let's try.

@bors-servo try

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11496#-:-KJBSQraXtPs1YlBP3Po:382119059)*

**Comment by nox** [comment]
The two other tests fail because they contradict the spec. I've filed https://github.com/w3c/web-platform-tests/issues/3103 about this.

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 1 of 1 files at r5, 1 of 1 files at r6, 9 of 9 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/range.rs, line 928 \[r7\]](https://reviewable.io:443/reviews/servo/servo/11496#-KJDG7O7YKFJPUMUFIrW:-KJDG7O7YKFJPUMUFIrX:1216518783) ([raw file](https://github.com/servo/servo/blob/44791a50052249f740941258f0e5c6adb4ac58b7/components/script/dom/range.rs#L928)):*
> ```Rust
>         // Step 4.
>         // Can be ignored because at the initialization all scripts in fragment node are supposed
>         // to be unmarked as "already started".
> ```

The test "`<script>`s should be run when appended to the document (but not before)" fails because I was wrong about this... You need to do a few things:

 * Change `HTMLScriptElement::mark_already_started` into `HTMLScriptElement::set_already_started`.
 * Traverse `fragment_node` (with `Node::traverse_preorder`) and call `script.set_already_started(false)` on all scripts in `fragment_node`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11496#-:-KJDHHiZmkHrpujOMhhL:244491485)*

**Comment by nox** [comment]
I'm wrong about the two other tests, they should pass too. I'm looking into why they fail.

<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> The two other tests fail because they contradict the spec. I've filed https://github.com/w3c/web-platform-tests/issues/3103 about this.
</details>


---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11496#-:-KJFsExr4dC1HGL0lbRQ:1739357072)*

**Comment by nox** [comment]
-S-awaiting-review +S-needs-code-changes 

Please also run the test `/domparsing/createContextualFragment.html` to check that the failing test about `<script>` now pass and update the expectations accordingly.

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/range.rs, line 929 \[r8\]](https://reviewable.io:443/reviews/servo/servo/11496#-KJIMdIvc5qCkk97F3P2:-KJIMdIvc5qCkk97F3P3:-1244770616) ([raw file](https://github.com/servo/servo/blob/e8db0713782d8788b9fad3e66e9ac21b8108753c/components/script/dom/range.rs#L929)):*
> ```Rust
>         // Step 4.
>         for node in fragment_node.upcast::<Node>().traverse_preorder() {
>             node.downcast::<HTMLScriptElement>().unwrap().set_already_started(false);
> ```

That's wrong. Not everything is a `HTMLScriptElement` in `fragment_node`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11496#-:-KJIMk5ZAvyOyg1VU4ut:-1758790669)*

**Comment by nox** [comment]
@bors-servo try

-S-awaiting-review 

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 4 of 4 files at r9, 3 of 3 files at r10, 12 of 12 files at r11, 12 of 12 files at r12.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11496#-:-KJIUaV1DNtVnDPrZ3FY:1386239646)*

**Comment by nox** [comment]
@bors-servo r+

I'll file a followup for the `<script>` issue.

**Pull request merged**